{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d02c83f8-05ee-46d0-990a-bd1940035841",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Then import the required libraries\n",
    "import importlib\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "348e6c47-f9b3-4674-8cb8-13c4deb51a27",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original columns: ['GeneSymbol', 'SingleValueType', 'TCGA-OL-A66H-01', 'TCGA-3C-AALK-01', 'TCGA-AC-A5EH-01', 'TCGA-EW-A2FW-01', 'TCGA-E9-A1R0-01', 'TCGA-AR-A1AJ-01', 'TCGA-AC-A62Y-01', 'TCGA-E9-A1QZ-01'] ...\n",
      "Column 'SingleValueType' has been removed successfully.\n",
      "CNV Matrix After Transpose and Header Fix: (785, 19277)\n",
      "\n",
      "DataFrame preview:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>RPS4Y2</th>\n",
       "      <th>RPS4Y1</th>\n",
       "      <th>TBL1Y</th>\n",
       "      <th>NLGN4Y</th>\n",
       "      <th>PRKY</th>\n",
       "      <th>PCDH11Y</th>\n",
       "      <th>TTTY20</th>\n",
       "      <th>GABRE</th>\n",
       "      <th>FTHL17</th>\n",
       "      <th>PORCN</th>\n",
       "      <th>...</th>\n",
       "      <th>EP300</th>\n",
       "      <th>DERL3</th>\n",
       "      <th>PLA2G3</th>\n",
       "      <th>TMPRSS6</th>\n",
       "      <th>MIR1281</th>\n",
       "      <th>GGT3P</th>\n",
       "      <th>UBE2L3</th>\n",
       "      <th>APOL5</th>\n",
       "      <th>SLC5A1</th>\n",
       "      <th>CDC42EP1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AAAU-01</th>\n",
       "      <td>0.528746</td>\n",
       "      <td>0.573773</td>\n",
       "      <td>0.440706</td>\n",
       "      <td>0.053398</td>\n",
       "      <td>0.117736</td>\n",
       "      <td>0.050879</td>\n",
       "      <td>0.51045</td>\n",
       "      <td>0.371732</td>\n",
       "      <td>0.877801</td>\n",
       "      <td>0.360481</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0974</td>\n",
       "      <td>0.056637</td>\n",
       "      <td>0.806611</td>\n",
       "      <td>0.868405</td>\n",
       "      <td>0.0974</td>\n",
       "      <td>0.874078</td>\n",
       "      <td>0.31182</td>\n",
       "      <td>0.931256</td>\n",
       "      <td>0.186356</td>\n",
       "      <td>0.017439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALI-01</th>\n",
       "      <td>0.651071</td>\n",
       "      <td>0.509548</td>\n",
       "      <td>0.41667</td>\n",
       "      <td>0.072804</td>\n",
       "      <td>0.806081</td>\n",
       "      <td>0.07175</td>\n",
       "      <td>0.485439</td>\n",
       "      <td>0.378231</td>\n",
       "      <td>0.757559</td>\n",
       "      <td>0.249911</td>\n",
       "      <td>...</td>\n",
       "      <td>0.060027</td>\n",
       "      <td>0.406407</td>\n",
       "      <td>0.78659</td>\n",
       "      <td>0.82704</td>\n",
       "      <td>0.060027</td>\n",
       "      <td>0.84627</td>\n",
       "      <td>0.78181</td>\n",
       "      <td>0.879644</td>\n",
       "      <td>0.899155</td>\n",
       "      <td>0.022864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALJ-01</th>\n",
       "      <td>0.697286</td>\n",
       "      <td>0.56616</td>\n",
       "      <td>0.453496</td>\n",
       "      <td>0.068645</td>\n",
       "      <td>0.782943</td>\n",
       "      <td>0.051312</td>\n",
       "      <td>0.469256</td>\n",
       "      <td>0.324052</td>\n",
       "      <td>0.627746</td>\n",
       "      <td>0.466564</td>\n",
       "      <td>...</td>\n",
       "      <td>0.072913</td>\n",
       "      <td>0.508306</td>\n",
       "      <td>0.787246</td>\n",
       "      <td>0.848229</td>\n",
       "      <td>0.072913</td>\n",
       "      <td>0.809162</td>\n",
       "      <td>0.409583</td>\n",
       "      <td>0.894666</td>\n",
       "      <td>0.738759</td>\n",
       "      <td>0.030299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALK-01</th>\n",
       "      <td>0.580953</td>\n",
       "      <td>0.575181</td>\n",
       "      <td>0.496185</td>\n",
       "      <td>0.327338</td>\n",
       "      <td>0.670718</td>\n",
       "      <td>0.242422</td>\n",
       "      <td>0.593454</td>\n",
       "      <td>0.527467</td>\n",
       "      <td>0.923862</td>\n",
       "      <td>0.408977</td>\n",
       "      <td>...</td>\n",
       "      <td>0.082873</td>\n",
       "      <td>0.569962</td>\n",
       "      <td>0.777228</td>\n",
       "      <td>0.841816</td>\n",
       "      <td>0.082873</td>\n",
       "      <td>0.836196</td>\n",
       "      <td>0.26714</td>\n",
       "      <td>0.937831</td>\n",
       "      <td>0.762713</td>\n",
       "      <td>0.026589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-4H-AAAK-01</th>\n",
       "      <td>0.474765</td>\n",
       "      <td>0.510691</td>\n",
       "      <td>0.440558</td>\n",
       "      <td>0.060931</td>\n",
       "      <td>0.642323</td>\n",
       "      <td>0.320909</td>\n",
       "      <td>0.454611</td>\n",
       "      <td>0.469725</td>\n",
       "      <td>0.916842</td>\n",
       "      <td>0.461385</td>\n",
       "      <td>...</td>\n",
       "      <td>0.078164</td>\n",
       "      <td>0.202899</td>\n",
       "      <td>0.696572</td>\n",
       "      <td>0.776501</td>\n",
       "      <td>0.078164</td>\n",
       "      <td>0.826303</td>\n",
       "      <td>0.242883</td>\n",
       "      <td>0.942489</td>\n",
       "      <td>0.7204</td>\n",
       "      <td>0.026004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19277 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GeneSymbol         RPS4Y2    RPS4Y1     TBL1Y    NLGN4Y      PRKY   PCDH11Y  \\\n",
       "TCGA-3C-AAAU-01  0.528746  0.573773  0.440706  0.053398  0.117736  0.050879   \n",
       "TCGA-3C-AALI-01  0.651071  0.509548   0.41667  0.072804  0.806081   0.07175   \n",
       "TCGA-3C-AALJ-01  0.697286   0.56616  0.453496  0.068645  0.782943  0.051312   \n",
       "TCGA-3C-AALK-01  0.580953  0.575181  0.496185  0.327338  0.670718  0.242422   \n",
       "TCGA-4H-AAAK-01  0.474765  0.510691  0.440558  0.060931  0.642323  0.320909   \n",
       "\n",
       "GeneSymbol         TTTY20     GABRE    FTHL17     PORCN  ...     EP300  \\\n",
       "TCGA-3C-AAAU-01   0.51045  0.371732  0.877801  0.360481  ...    0.0974   \n",
       "TCGA-3C-AALI-01  0.485439  0.378231  0.757559  0.249911  ...  0.060027   \n",
       "TCGA-3C-AALJ-01  0.469256  0.324052  0.627746  0.466564  ...  0.072913   \n",
       "TCGA-3C-AALK-01  0.593454  0.527467  0.923862  0.408977  ...  0.082873   \n",
       "TCGA-4H-AAAK-01  0.454611  0.469725  0.916842  0.461385  ...  0.078164   \n",
       "\n",
       "GeneSymbol          DERL3    PLA2G3   TMPRSS6   MIR1281     GGT3P    UBE2L3  \\\n",
       "TCGA-3C-AAAU-01  0.056637  0.806611  0.868405    0.0974  0.874078   0.31182   \n",
       "TCGA-3C-AALI-01  0.406407   0.78659   0.82704  0.060027   0.84627   0.78181   \n",
       "TCGA-3C-AALJ-01  0.508306  0.787246  0.848229  0.072913  0.809162  0.409583   \n",
       "TCGA-3C-AALK-01  0.569962  0.777228  0.841816  0.082873  0.836196   0.26714   \n",
       "TCGA-4H-AAAK-01  0.202899  0.696572  0.776501  0.078164  0.826303  0.242883   \n",
       "\n",
       "GeneSymbol          APOL5    SLC5A1  CDC42EP1  \n",
       "TCGA-3C-AAAU-01  0.931256  0.186356  0.017439  \n",
       "TCGA-3C-AALI-01  0.879644  0.899155  0.022864  \n",
       "TCGA-3C-AALJ-01  0.894666  0.738759  0.030299  \n",
       "TCGA-3C-AALK-01  0.937831  0.762713  0.026589  \n",
       "TCGA-4H-AAAK-01  0.942489    0.7204  0.026004  \n",
       "\n",
       "[5 rows x 19277 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seed = 0\n",
    "np.random.seed(seed)\n",
    "\n",
    "df = pd.read_csv('gene_level_methylation.csv')\n",
    "\n",
    "print(\"Original columns:\", df.columns.tolist()[:10], \"...\")\n",
    "\n",
    "# Remove the \"SingleValueType\" column if it exists\n",
    "if \"SingleValueType\" in df.columns:\n",
    "    df = df.drop(columns=[\"SingleValueType\"])\n",
    "    print(\"Column 'SingleValueType' has been removed successfully.\")\n",
    "else:\n",
    "    print(\"Column 'SingleValueType' does not exist in the DataFrame.\")\n",
    "\n",
    "# Sort columns and transpose\n",
    "df = df.reindex(sorted(df.columns), axis=1)\n",
    "df = df.T\n",
    "\n",
    "# Set first row (originally \"GeneSymbol\") as new column names\n",
    "df.columns = df.iloc[0]  # first row becomes column headers\n",
    "df = df.drop(df.index[0])  # drop the first row\n",
    "\n",
    "print(\"CNV Matrix After Transpose and Header Fix:\", df.shape)\n",
    "print(\"\\nDataFrame preview:\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cef509c0-97a6-4354-9ec4-d5ac88c1b3f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clinical Data Before: (1247, 202)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AJCC_Stage_nature2012</th>\n",
       "      <th>Age_at_Initial_Pathologic_Diagnosis_nature2012</th>\n",
       "      <th>CN_Clusters_nature2012</th>\n",
       "      <th>Converted_Stage_nature2012</th>\n",
       "      <th>Days_to_Date_of_Last_Contact_nature2012</th>\n",
       "      <th>Days_to_date_of_Death_nature2012</th>\n",
       "      <th>ER_Status_nature2012</th>\n",
       "      <th>Gender_nature2012</th>\n",
       "      <th>HER2_Final_Status_nature2012</th>\n",
       "      <th>Integrated_Clusters_no_exp__nature2012</th>\n",
       "      <th>...</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_mutation_wustl_gene</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_miRNA_GA</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_percentile</th>\n",
       "      <th>_GENOMIC_ID_data/public/TCGA/BRCA/miRNA_GA_gene</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_gistic2thd</th>\n",
       "      <th>_GENOMIC_ID_data/public/TCGA/BRCA/miRNA_HiSeq_gene</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_G4502A_07_3</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_gistic2</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_PDMarray</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SB-01</th>\n",
       "      <td>Stage I</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>259.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-A1-A0SB-01A-11D-A142-09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a2405d64-34eb-4915-abf7-8530151d5cb0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A1-A0SB-01A-11D-A141-01</td>\n",
       "      <td>TCGA-A1-A0SB-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a2405d64-34eb-4915-abf7-8530151d5cb0</td>\n",
       "      <td>TCGA-A1-A0SB-01A-11D-A141-01</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SD-01</th>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>59.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>437.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-A1-A0SD-01A-11D-A10Y-09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15bad71d-3031-413b-9e8d-6426ae5dfbea</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A1-A0SD-01A-11D-A111-01</td>\n",
       "      <td>TCGA-A1-A0SD-01</td>\n",
       "      <td>TCGA-A1-A0SD-01A-11R-A115-07</td>\n",
       "      <td>15bad71d-3031-413b-9e8d-6426ae5dfbea</td>\n",
       "      <td>TCGA-A1-A0SD-01A-11D-A111-01</td>\n",
       "      <td>TCGA-A1-A0SD-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SE-01</th>\n",
       "      <td>Stage I</td>\n",
       "      <td>56.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>1320.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-A1-A0SE-01A-11D-A099-09</td>\n",
       "      <td>TCGA-A1-A0SE-01</td>\n",
       "      <td>a998e0ce-9248-460f-aabc-2dad452a1ff9</td>\n",
       "      <td>TCGA-A1-A0SE-01</td>\n",
       "      <td>TCGA-A1-A0SE-01A-11D-A087-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A1-A0SE-01A-11R-A084-07</td>\n",
       "      <td>a998e0ce-9248-460f-aabc-2dad452a1ff9</td>\n",
       "      <td>TCGA-A1-A0SE-01A-11D-A087-01</td>\n",
       "      <td>TCGA-A1-A0SE-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SF-01</th>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>54.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>1463.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-A1-A0SF-01A-11D-A142-09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28089e15-5e2c-4e83-ba6c-62b3cb40e431</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A1-A0SF-01A-11D-A141-01</td>\n",
       "      <td>TCGA-A1-A0SF-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28089e15-5e2c-4e83-ba6c-62b3cb40e431</td>\n",
       "      <td>TCGA-A1-A0SF-01A-11D-A141-01</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SG-01</th>\n",
       "      <td>Stage IIB</td>\n",
       "      <td>61.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Stage IIB</td>\n",
       "      <td>433.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-A1-A0SG-01A-11D-A142-09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0df6b948-367d-4951-9d98-d3bebabff63e</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A1-A0SG-01A-11D-A141-01</td>\n",
       "      <td>TCGA-A1-A0SG-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0df6b948-367d-4951-9d98-d3bebabff63e</td>\n",
       "      <td>TCGA-A1-A0SG-01A-11D-A141-01</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                AJCC_Stage_nature2012  \\\n",
       "TCGA-A1-A0SB-01               Stage I   \n",
       "TCGA-A1-A0SD-01             Stage IIA   \n",
       "TCGA-A1-A0SE-01               Stage I   \n",
       "TCGA-A1-A0SF-01             Stage IIA   \n",
       "TCGA-A1-A0SG-01             Stage IIB   \n",
       "\n",
       "                 Age_at_Initial_Pathologic_Diagnosis_nature2012  \\\n",
       "TCGA-A1-A0SB-01                                            70.0   \n",
       "TCGA-A1-A0SD-01                                            59.0   \n",
       "TCGA-A1-A0SE-01                                            56.0   \n",
       "TCGA-A1-A0SF-01                                            54.0   \n",
       "TCGA-A1-A0SG-01                                            61.0   \n",
       "\n",
       "                 CN_Clusters_nature2012 Converted_Stage_nature2012  \\\n",
       "TCGA-A1-A0SB-01                     1.0                    Stage I   \n",
       "TCGA-A1-A0SD-01                     2.0                  Stage IIA   \n",
       "TCGA-A1-A0SE-01                     2.0                    Stage I   \n",
       "TCGA-A1-A0SF-01                     3.0                  Stage IIA   \n",
       "TCGA-A1-A0SG-01                     4.0                  Stage IIB   \n",
       "\n",
       "                 Days_to_Date_of_Last_Contact_nature2012  \\\n",
       "TCGA-A1-A0SB-01                                    259.0   \n",
       "TCGA-A1-A0SD-01                                    437.0   \n",
       "TCGA-A1-A0SE-01                                   1320.0   \n",
       "TCGA-A1-A0SF-01                                   1463.0   \n",
       "TCGA-A1-A0SG-01                                    433.0   \n",
       "\n",
       "                 Days_to_date_of_Death_nature2012 ER_Status_nature2012  \\\n",
       "TCGA-A1-A0SB-01                               NaN             Positive   \n",
       "TCGA-A1-A0SD-01                               NaN             Positive   \n",
       "TCGA-A1-A0SE-01                               NaN             Positive   \n",
       "TCGA-A1-A0SF-01                               NaN             Positive   \n",
       "TCGA-A1-A0SG-01                               NaN             Positive   \n",
       "\n",
       "                Gender_nature2012 HER2_Final_Status_nature2012  \\\n",
       "TCGA-A1-A0SB-01            FEMALE                     Negative   \n",
       "TCGA-A1-A0SD-01            FEMALE                     Negative   \n",
       "TCGA-A1-A0SE-01            FEMALE                     Negative   \n",
       "TCGA-A1-A0SF-01            FEMALE                     Negative   \n",
       "TCGA-A1-A0SG-01            FEMALE                     Negative   \n",
       "\n",
       "                 Integrated_Clusters_no_exp__nature2012  ...  \\\n",
       "TCGA-A1-A0SB-01                                     NaN  ...   \n",
       "TCGA-A1-A0SD-01                                     NaN  ...   \n",
       "TCGA-A1-A0SE-01                                     NaN  ...   \n",
       "TCGA-A1-A0SF-01                                     NaN  ...   \n",
       "TCGA-A1-A0SG-01                                     NaN  ...   \n",
       "\n",
       "                 _GENOMIC_ID_TCGA_BRCA_mutation_wustl_gene  \\\n",
       "TCGA-A1-A0SB-01               TCGA-A1-A0SB-01A-11D-A142-09   \n",
       "TCGA-A1-A0SD-01               TCGA-A1-A0SD-01A-11D-A10Y-09   \n",
       "TCGA-A1-A0SE-01               TCGA-A1-A0SE-01A-11D-A099-09   \n",
       "TCGA-A1-A0SF-01               TCGA-A1-A0SF-01A-11D-A142-09   \n",
       "TCGA-A1-A0SG-01               TCGA-A1-A0SG-01A-11D-A142-09   \n",
       "\n",
       "                 _GENOMIC_ID_TCGA_BRCA_miRNA_GA  \\\n",
       "TCGA-A1-A0SB-01                             NaN   \n",
       "TCGA-A1-A0SD-01                             NaN   \n",
       "TCGA-A1-A0SE-01                 TCGA-A1-A0SE-01   \n",
       "TCGA-A1-A0SF-01                             NaN   \n",
       "TCGA-A1-A0SG-01                             NaN   \n",
       "\n",
       "                _GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_percentile  \\\n",
       "TCGA-A1-A0SB-01         a2405d64-34eb-4915-abf7-8530151d5cb0   \n",
       "TCGA-A1-A0SD-01         15bad71d-3031-413b-9e8d-6426ae5dfbea   \n",
       "TCGA-A1-A0SE-01         a998e0ce-9248-460f-aabc-2dad452a1ff9   \n",
       "TCGA-A1-A0SF-01         28089e15-5e2c-4e83-ba6c-62b3cb40e431   \n",
       "TCGA-A1-A0SG-01         0df6b948-367d-4951-9d98-d3bebabff63e   \n",
       "\n",
       "                _GENOMIC_ID_data/public/TCGA/BRCA/miRNA_GA_gene  \\\n",
       "TCGA-A1-A0SB-01                                             NaN   \n",
       "TCGA-A1-A0SD-01                                             NaN   \n",
       "TCGA-A1-A0SE-01                                 TCGA-A1-A0SE-01   \n",
       "TCGA-A1-A0SF-01                                             NaN   \n",
       "TCGA-A1-A0SG-01                                             NaN   \n",
       "\n",
       "                _GENOMIC_ID_TCGA_BRCA_gistic2thd  \\\n",
       "TCGA-A1-A0SB-01     TCGA-A1-A0SB-01A-11D-A141-01   \n",
       "TCGA-A1-A0SD-01     TCGA-A1-A0SD-01A-11D-A111-01   \n",
       "TCGA-A1-A0SE-01     TCGA-A1-A0SE-01A-11D-A087-01   \n",
       "TCGA-A1-A0SF-01     TCGA-A1-A0SF-01A-11D-A141-01   \n",
       "TCGA-A1-A0SG-01     TCGA-A1-A0SG-01A-11D-A141-01   \n",
       "\n",
       "                _GENOMIC_ID_data/public/TCGA/BRCA/miRNA_HiSeq_gene  \\\n",
       "TCGA-A1-A0SB-01                                    TCGA-A1-A0SB-01   \n",
       "TCGA-A1-A0SD-01                                    TCGA-A1-A0SD-01   \n",
       "TCGA-A1-A0SE-01                                                NaN   \n",
       "TCGA-A1-A0SF-01                                    TCGA-A1-A0SF-01   \n",
       "TCGA-A1-A0SG-01                                    TCGA-A1-A0SG-01   \n",
       "\n",
       "                 _GENOMIC_ID_TCGA_BRCA_G4502A_07_3  \\\n",
       "TCGA-A1-A0SB-01                                NaN   \n",
       "TCGA-A1-A0SD-01       TCGA-A1-A0SD-01A-11R-A115-07   \n",
       "TCGA-A1-A0SE-01       TCGA-A1-A0SE-01A-11R-A084-07   \n",
       "TCGA-A1-A0SF-01                                NaN   \n",
       "TCGA-A1-A0SG-01                                NaN   \n",
       "\n",
       "                    _GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2  \\\n",
       "TCGA-A1-A0SB-01  a2405d64-34eb-4915-abf7-8530151d5cb0   \n",
       "TCGA-A1-A0SD-01  15bad71d-3031-413b-9e8d-6426ae5dfbea   \n",
       "TCGA-A1-A0SE-01  a998e0ce-9248-460f-aabc-2dad452a1ff9   \n",
       "TCGA-A1-A0SF-01  28089e15-5e2c-4e83-ba6c-62b3cb40e431   \n",
       "TCGA-A1-A0SG-01  0df6b948-367d-4951-9d98-d3bebabff63e   \n",
       "\n",
       "                _GENOMIC_ID_TCGA_BRCA_gistic2 _GENOMIC_ID_TCGA_BRCA_PDMarray  \n",
       "TCGA-A1-A0SB-01  TCGA-A1-A0SB-01A-11D-A141-01                            NaN  \n",
       "TCGA-A1-A0SD-01  TCGA-A1-A0SD-01A-11D-A111-01                TCGA-A1-A0SD-01  \n",
       "TCGA-A1-A0SE-01  TCGA-A1-A0SE-01A-11D-A087-01                TCGA-A1-A0SE-01  \n",
       "TCGA-A1-A0SF-01  TCGA-A1-A0SF-01A-11D-A141-01                            NaN  \n",
       "TCGA-A1-A0SG-01  TCGA-A1-A0SG-01A-11D-A141-01                            NaN  \n",
       "\n",
       "[5 rows x 202 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2=pd.read_csv('BRCA_clinicalMatrix.gz',compression='gzip',sep='\\t',index_col=0)\n",
    "df2.index = pd.Series({x: x.replace('-', '-') for x in df2.index})\n",
    "print(\"Clinical Data Before:\", df2.shape)\n",
    "k=df2.columns.get_loc('PAM50Call_RNAseq')\n",
    "df2=df2[df2.iloc[:,k].isna()==False]\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e78c03fd-abfd-42c0-a2db-3e67c7c332cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data After: (539, 19277)\n",
      "Clinical Data After: (539, 202)\n",
      "CNV and Clinical Data corresponds to same Patients:  True\n",
      "Labels ['Basal' 'Her2' 'LumA' 'LumB' 'Normal'] [ 86  31 279 126  17]\n"
     ]
    }
   ],
   "source": [
    "commonIndexes=np.sort(list(set(df2.index).intersection(set(df.index)))) \n",
    "df=df.loc[commonIndexes] \n",
    "df2=df2.loc[commonIndexes] \n",
    "print(\"Data After:\", df.shape) \n",
    "print(\"Clinical Data After:\", df2.shape) \n",
    "print(\"CNV and Clinical Data corresponds to same Patients: \", False if False in (df.index==df2.index) else True) \n",
    "unique_elements, counts_elements = np.unique(df2.iloc[:,k], return_counts=True) \n",
    "print(\"Labels\", unique_elements, counts_elements)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "682a6675-d9e9-423a-a6b1-b2bc2a75e3c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(539,)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "original_feature_names = df.columns.tolist()\n",
    "X=np.array(df.values, dtype=float)\n",
    "X.shape\n",
    "Y=np.array(df2.values[:,k])\n",
    "Y.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "978a01a9-a906-4d4a-8983-ae3e34710ad3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import StratifiedKFold\n",
    "\n",
    "n_splits = 5\n",
    "skf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8a3c1439-e5c0-4c71-8a4c-f9903793cdb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((432, 19277), (107, 19277))"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "folds = list(skf.split(X, Y))\n",
    "\n",
    "# Select fold 4\n",
    "train_idx, test_idx = folds[4]\n",
    "X_train, X_test = X[train_idx], X[test_idx]\n",
    "Y_train, Y_test = Y[train_idx], Y[test_idx]\n",
    "\n",
    "X_train.shape, X_test.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9494c4ea-d833-4edf-9fd9-d80ca7b91b14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train label distribution:\n",
      " Label\n",
      "LumA      223\n",
      "LumB      101\n",
      "Basal      69\n",
      "Her2       25\n",
      "Normal     14\n",
      "Name: count, dtype: int64\n",
      "Test label distribution:\n",
      " Label\n",
      "LumA      56\n",
      "LumB      25\n",
      "Basal     17\n",
      "Her2       6\n",
      "Normal     3\n",
      "Name: count, dtype: int64\n",
      "Train/test CSVs saved successfully under 'TCGA Dataset Splits/' folder.\n"
     ]
    }
   ],
   "source": [
    "# Convert back to DataFrames for clarity and labeled saving\n",
    "X_train_df = pd.DataFrame(X_train, columns=original_feature_names, index=np.arange(len(X_train)))\n",
    "X_test_df = pd.DataFrame(X_test, columns=original_feature_names, index=np.arange(len(X_test)))\n",
    "Y_train_df = pd.DataFrame(Y_train, columns=[\"Label\"], index=np.arange(len(Y_train)))\n",
    "Y_test_df = pd.DataFrame(Y_test, columns=[\"Label\"], index=np.arange(len(Y_test)))\n",
    "\n",
    "print(\"Train label distribution:\\n\", Y_train_df[\"Label\"].value_counts())\n",
    "print(\"Test label distribution:\\n\", Y_test_df[\"Label\"].value_counts())\n",
    "\n",
    "os.makedirs(\"splits\", exist_ok=True)\n",
    "\n",
    "# Save to CSV files\n",
    "X_train_df.to_csv(\"Datasets/TCGA Dataset Splits/X_train.csv\", index=False)\n",
    "X_test_df.to_csv(\"Datasets/TCGA Dataset Splits/X_test.csv\", index=False)\n",
    "Y_train_df.to_csv(\"Datasets/TCGA Dataset Splits/Y_train.csv\", index=False)\n",
    "Y_test_df.to_csv(\"Datasets/TCGA Dataset Splits/Y_test.csv\", index=False)\n",
    "\n",
    "print(\"Train/test CSVs saved successfully under 'TCGA Dataset Splits/' folder.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a79654f-7c95-4eeb-b107-4587ff9d7ce5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
