{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d02c83f8-05ee-46d0-990a-bd1940035841",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Then import the required libraries\n",
    "import importlib\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "348e6c47-f9b3-4674-8cb8-13c4deb51a27",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original columns: ['GeneSymbol', 'SingleValueType', 'TCGA-OL-A66H-01', 'TCGA-3C-AALK-01', 'TCGA-AC-A5EH-01', 'TCGA-EW-A2FW-01', 'TCGA-E9-A1R0-01', 'TCGA-AR-A1AJ-01', 'TCGA-AC-A62Y-01', 'TCGA-E9-A1QZ-01'] ...\n",
      "Column 'SingleValueType' has been removed successfully.\n",
      "CNV Matrix After Transpose and Header Fix: (785, 5351)\n",
      "\n",
      "DataFrame preview:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>NLGN4Y</th>\n",
       "      <th>HS6ST2</th>\n",
       "      <th>ARMCX3</th>\n",
       "      <th>GNL3L</th>\n",
       "      <th>BEX5</th>\n",
       "      <th>TSPYL2</th>\n",
       "      <th>PNMA3</th>\n",
       "      <th>PNCK</th>\n",
       "      <th>CLCN5</th>\n",
       "      <th>DACH2</th>\n",
       "      <th>...</th>\n",
       "      <th>TIMP3</th>\n",
       "      <th>RHBDD3</th>\n",
       "      <th>C22orf24</th>\n",
       "      <th>TFIP11</th>\n",
       "      <th>ADORA2A</th>\n",
       "      <th>CCDC134</th>\n",
       "      <th>TMPRSS6</th>\n",
       "      <th>POLDIP3</th>\n",
       "      <th>YWHAH</th>\n",
       "      <th>RNU12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AAAU-01</th>\n",
       "      <td>0.01263</td>\n",
       "      <td>0.929187</td>\n",
       "      <td>0.910775</td>\n",
       "      <td>0.029008</td>\n",
       "      <td>0.10617</td>\n",
       "      <td>0.16089</td>\n",
       "      <td>0.862272</td>\n",
       "      <td>0.343161</td>\n",
       "      <td>0.606573</td>\n",
       "      <td>0.096672</td>\n",
       "      <td>...</td>\n",
       "      <td>0.409331</td>\n",
       "      <td>0.024264</td>\n",
       "      <td>0.063856</td>\n",
       "      <td>0.029824</td>\n",
       "      <td>0.982844</td>\n",
       "      <td>0.017207</td>\n",
       "      <td>0.875728</td>\n",
       "      <td>0.040087</td>\n",
       "      <td>0.063856</td>\n",
       "      <td>0.040087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALI-01</th>\n",
       "      <td>0.027041</td>\n",
       "      <td>0.845574</td>\n",
       "      <td>0.837298</td>\n",
       "      <td>0.021677</td>\n",
       "      <td>0.194219</td>\n",
       "      <td>0.071764</td>\n",
       "      <td>0.860443</td>\n",
       "      <td>0.30153</td>\n",
       "      <td>0.503716</td>\n",
       "      <td>0.188203</td>\n",
       "      <td>...</td>\n",
       "      <td>0.735495</td>\n",
       "      <td>0.021732</td>\n",
       "      <td>0.078717</td>\n",
       "      <td>0.044712</td>\n",
       "      <td>0.965711</td>\n",
       "      <td>0.022468</td>\n",
       "      <td>0.905563</td>\n",
       "      <td>0.036102</td>\n",
       "      <td>0.078717</td>\n",
       "      <td>0.036102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALJ-01</th>\n",
       "      <td>0.01927</td>\n",
       "      <td>0.80953</td>\n",
       "      <td>0.802017</td>\n",
       "      <td>0.114992</td>\n",
       "      <td>0.242886</td>\n",
       "      <td>0.204084</td>\n",
       "      <td>0.715122</td>\n",
       "      <td>0.579739</td>\n",
       "      <td>0.591016</td>\n",
       "      <td>0.344009</td>\n",
       "      <td>...</td>\n",
       "      <td>0.489235</td>\n",
       "      <td>0.025593</td>\n",
       "      <td>0.073055</td>\n",
       "      <td>0.048104</td>\n",
       "      <td>0.958028</td>\n",
       "      <td>0.018441</td>\n",
       "      <td>0.891938</td>\n",
       "      <td>0.03992</td>\n",
       "      <td>0.073055</td>\n",
       "      <td>0.03992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALK-01</th>\n",
       "      <td>0.333264</td>\n",
       "      <td>0.682301</td>\n",
       "      <td>0.790784</td>\n",
       "      <td>0.25185</td>\n",
       "      <td>0.169045</td>\n",
       "      <td>0.173549</td>\n",
       "      <td>0.847816</td>\n",
       "      <td>0.36839</td>\n",
       "      <td>0.757681</td>\n",
       "      <td>0.342259</td>\n",
       "      <td>...</td>\n",
       "      <td>0.574279</td>\n",
       "      <td>0.021366</td>\n",
       "      <td>0.072598</td>\n",
       "      <td>0.038266</td>\n",
       "      <td>0.97526</td>\n",
       "      <td>0.018463</td>\n",
       "      <td>0.881912</td>\n",
       "      <td>0.045846</td>\n",
       "      <td>0.072598</td>\n",
       "      <td>0.045846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-4H-AAAK-01</th>\n",
       "      <td>0.03447</td>\n",
       "      <td>0.662762</td>\n",
       "      <td>0.694734</td>\n",
       "      <td>0.022959</td>\n",
       "      <td>0.265914</td>\n",
       "      <td>0.200786</td>\n",
       "      <td>0.759122</td>\n",
       "      <td>0.502064</td>\n",
       "      <td>0.747459</td>\n",
       "      <td>0.365961</td>\n",
       "      <td>...</td>\n",
       "      <td>0.411259</td>\n",
       "      <td>0.022462</td>\n",
       "      <td>0.091905</td>\n",
       "      <td>0.038059</td>\n",
       "      <td>0.984475</td>\n",
       "      <td>0.025709</td>\n",
       "      <td>0.874579</td>\n",
       "      <td>0.044324</td>\n",
       "      <td>0.091905</td>\n",
       "      <td>0.044324</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 5351 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GeneSymbol         NLGN4Y    HS6ST2    ARMCX3     GNL3L      BEX5    TSPYL2  \\\n",
       "TCGA-3C-AAAU-01   0.01263  0.929187  0.910775  0.029008   0.10617   0.16089   \n",
       "TCGA-3C-AALI-01  0.027041  0.845574  0.837298  0.021677  0.194219  0.071764   \n",
       "TCGA-3C-AALJ-01   0.01927   0.80953  0.802017  0.114992  0.242886  0.204084   \n",
       "TCGA-3C-AALK-01  0.333264  0.682301  0.790784   0.25185  0.169045  0.173549   \n",
       "TCGA-4H-AAAK-01   0.03447  0.662762  0.694734  0.022959  0.265914  0.200786   \n",
       "\n",
       "GeneSymbol          PNMA3      PNCK     CLCN5     DACH2  ...     TIMP3  \\\n",
       "TCGA-3C-AAAU-01  0.862272  0.343161  0.606573  0.096672  ...  0.409331   \n",
       "TCGA-3C-AALI-01  0.860443   0.30153  0.503716  0.188203  ...  0.735495   \n",
       "TCGA-3C-AALJ-01  0.715122  0.579739  0.591016  0.344009  ...  0.489235   \n",
       "TCGA-3C-AALK-01  0.847816   0.36839  0.757681  0.342259  ...  0.574279   \n",
       "TCGA-4H-AAAK-01  0.759122  0.502064  0.747459  0.365961  ...  0.411259   \n",
       "\n",
       "GeneSymbol         RHBDD3  C22orf24    TFIP11   ADORA2A   CCDC134   TMPRSS6  \\\n",
       "TCGA-3C-AAAU-01  0.024264  0.063856  0.029824  0.982844  0.017207  0.875728   \n",
       "TCGA-3C-AALI-01  0.021732  0.078717  0.044712  0.965711  0.022468  0.905563   \n",
       "TCGA-3C-AALJ-01  0.025593  0.073055  0.048104  0.958028  0.018441  0.891938   \n",
       "TCGA-3C-AALK-01  0.021366  0.072598  0.038266   0.97526  0.018463  0.881912   \n",
       "TCGA-4H-AAAK-01  0.022462  0.091905  0.038059  0.984475  0.025709  0.874579   \n",
       "\n",
       "GeneSymbol        POLDIP3     YWHAH     RNU12  \n",
       "TCGA-3C-AAAU-01  0.040087  0.063856  0.040087  \n",
       "TCGA-3C-AALI-01  0.036102  0.078717  0.036102  \n",
       "TCGA-3C-AALJ-01   0.03992  0.073055   0.03992  \n",
       "TCGA-3C-AALK-01  0.045846  0.072598  0.045846  \n",
       "TCGA-4H-AAAK-01  0.044324  0.091905  0.044324  \n",
       "\n",
       "[5 rows x 5351 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seed = 0\n",
    "np.random.seed(seed)\n",
    "\n",
    "df = pd.read_csv('gene_level_methylation.csv')\n",
    "\n",
    "print(\"Original columns:\", df.columns.tolist()[:10], \"...\")\n",
    "\n",
    "# Remove the \"SingleValueType\" column if it exists\n",
    "if \"SingleValueType\" in df.columns:\n",
    "    df = df.drop(columns=[\"SingleValueType\"])\n",
    "    print(\"Column 'SingleValueType' has been removed successfully.\")\n",
    "else:\n",
    "    print(\"Column 'SingleValueType' does not exist in the DataFrame.\")\n",
    "\n",
    "# Sort columns and transpose\n",
    "df = df.reindex(sorted(df.columns), axis=1)\n",
    "df = df.T\n",
    "\n",
    "# Set first row (originally \"GeneSymbol\") as new column names\n",
    "df.columns = df.iloc[0]  # first row becomes column headers\n",
    "df = df.drop(df.index[0])  # drop the first row\n",
    "\n",
    "print(\"CNV Matrix After Transpose and Header Fix:\", df.shape)\n",
    "print(\"\\nDataFrame preview:\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cef509c0-97a6-4354-9ec4-d5ac88c1b3f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clinical Data Before: (1247, 202)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AJCC_Stage_nature2012</th>\n",
       "      <th>Age_at_Initial_Pathologic_Diagnosis_nature2012</th>\n",
       "      <th>CN_Clusters_nature2012</th>\n",
       "      <th>Converted_Stage_nature2012</th>\n",
       "      <th>Days_to_Date_of_Last_Contact_nature2012</th>\n",
       "      <th>Days_to_date_of_Death_nature2012</th>\n",
       "      <th>ER_Status_nature2012</th>\n",
       "      <th>Gender_nature2012</th>\n",
       "      <th>HER2_Final_Status_nature2012</th>\n",
       "      <th>Integrated_Clusters_no_exp__nature2012</th>\n",
       "      <th>...</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_mutation_wustl_gene</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_miRNA_GA</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_percentile</th>\n",
       "      <th>_GENOMIC_ID_data/public/TCGA/BRCA/miRNA_GA_gene</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_gistic2thd</th>\n",
       "      <th>_GENOMIC_ID_data/public/TCGA/BRCA/miRNA_HiSeq_gene</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_G4502A_07_3</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_gistic2</th>\n",
       "      <th>_GENOMIC_ID_TCGA_BRCA_PDMarray</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SB-01</th>\n",
       "      <td>Stage I</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>259.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-A1-A0SB-01A-11D-A142-09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a2405d64-34eb-4915-abf7-8530151d5cb0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A1-A0SB-01A-11D-A141-01</td>\n",
       "      <td>TCGA-A1-A0SB-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a2405d64-34eb-4915-abf7-8530151d5cb0</td>\n",
       "      <td>TCGA-A1-A0SB-01A-11D-A141-01</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SD-01</th>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>59.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>437.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-A1-A0SD-01A-11D-A10Y-09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15bad71d-3031-413b-9e8d-6426ae5dfbea</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A1-A0SD-01A-11D-A111-01</td>\n",
       "      <td>TCGA-A1-A0SD-01</td>\n",
       "      <td>TCGA-A1-A0SD-01A-11R-A115-07</td>\n",
       "      <td>15bad71d-3031-413b-9e8d-6426ae5dfbea</td>\n",
       "      <td>TCGA-A1-A0SD-01A-11D-A111-01</td>\n",
       "      <td>TCGA-A1-A0SD-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SE-01</th>\n",
       "      <td>Stage I</td>\n",
       "      <td>56.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>1320.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-A1-A0SE-01A-11D-A099-09</td>\n",
       "      <td>TCGA-A1-A0SE-01</td>\n",
       "      <td>a998e0ce-9248-460f-aabc-2dad452a1ff9</td>\n",
       "      <td>TCGA-A1-A0SE-01</td>\n",
       "      <td>TCGA-A1-A0SE-01A-11D-A087-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A1-A0SE-01A-11R-A084-07</td>\n",
       "      <td>a998e0ce-9248-460f-aabc-2dad452a1ff9</td>\n",
       "      <td>TCGA-A1-A0SE-01A-11D-A087-01</td>\n",
       "      <td>TCGA-A1-A0SE-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SF-01</th>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>54.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>1463.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-A1-A0SF-01A-11D-A142-09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28089e15-5e2c-4e83-ba6c-62b3cb40e431</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A1-A0SF-01A-11D-A141-01</td>\n",
       "      <td>TCGA-A1-A0SF-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28089e15-5e2c-4e83-ba6c-62b3cb40e431</td>\n",
       "      <td>TCGA-A1-A0SF-01A-11D-A141-01</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A1-A0SG-01</th>\n",
       "      <td>Stage IIB</td>\n",
       "      <td>61.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Stage IIB</td>\n",
       "      <td>433.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Positive</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>Negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>TCGA-A1-A0SG-01A-11D-A142-09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0df6b948-367d-4951-9d98-d3bebabff63e</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A1-A0SG-01A-11D-A141-01</td>\n",
       "      <td>TCGA-A1-A0SG-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0df6b948-367d-4951-9d98-d3bebabff63e</td>\n",
       "      <td>TCGA-A1-A0SG-01A-11D-A141-01</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 202 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                AJCC_Stage_nature2012  \\\n",
       "TCGA-A1-A0SB-01               Stage I   \n",
       "TCGA-A1-A0SD-01             Stage IIA   \n",
       "TCGA-A1-A0SE-01               Stage I   \n",
       "TCGA-A1-A0SF-01             Stage IIA   \n",
       "TCGA-A1-A0SG-01             Stage IIB   \n",
       "\n",
       "                 Age_at_Initial_Pathologic_Diagnosis_nature2012  \\\n",
       "TCGA-A1-A0SB-01                                            70.0   \n",
       "TCGA-A1-A0SD-01                                            59.0   \n",
       "TCGA-A1-A0SE-01                                            56.0   \n",
       "TCGA-A1-A0SF-01                                            54.0   \n",
       "TCGA-A1-A0SG-01                                            61.0   \n",
       "\n",
       "                 CN_Clusters_nature2012 Converted_Stage_nature2012  \\\n",
       "TCGA-A1-A0SB-01                     1.0                    Stage I   \n",
       "TCGA-A1-A0SD-01                     2.0                  Stage IIA   \n",
       "TCGA-A1-A0SE-01                     2.0                    Stage I   \n",
       "TCGA-A1-A0SF-01                     3.0                  Stage IIA   \n",
       "TCGA-A1-A0SG-01                     4.0                  Stage IIB   \n",
       "\n",
       "                 Days_to_Date_of_Last_Contact_nature2012  \\\n",
       "TCGA-A1-A0SB-01                                    259.0   \n",
       "TCGA-A1-A0SD-01                                    437.0   \n",
       "TCGA-A1-A0SE-01                                   1320.0   \n",
       "TCGA-A1-A0SF-01                                   1463.0   \n",
       "TCGA-A1-A0SG-01                                    433.0   \n",
       "\n",
       "                 Days_to_date_of_Death_nature2012 ER_Status_nature2012  \\\n",
       "TCGA-A1-A0SB-01                               NaN             Positive   \n",
       "TCGA-A1-A0SD-01                               NaN             Positive   \n",
       "TCGA-A1-A0SE-01                               NaN             Positive   \n",
       "TCGA-A1-A0SF-01                               NaN             Positive   \n",
       "TCGA-A1-A0SG-01                               NaN             Positive   \n",
       "\n",
       "                Gender_nature2012 HER2_Final_Status_nature2012  \\\n",
       "TCGA-A1-A0SB-01            FEMALE                     Negative   \n",
       "TCGA-A1-A0SD-01            FEMALE                     Negative   \n",
       "TCGA-A1-A0SE-01            FEMALE                     Negative   \n",
       "TCGA-A1-A0SF-01            FEMALE                     Negative   \n",
       "TCGA-A1-A0SG-01            FEMALE                     Negative   \n",
       "\n",
       "                 Integrated_Clusters_no_exp__nature2012  ...  \\\n",
       "TCGA-A1-A0SB-01                                     NaN  ...   \n",
       "TCGA-A1-A0SD-01                                     NaN  ...   \n",
       "TCGA-A1-A0SE-01                                     NaN  ...   \n",
       "TCGA-A1-A0SF-01                                     NaN  ...   \n",
       "TCGA-A1-A0SG-01                                     NaN  ...   \n",
       "\n",
       "                 _GENOMIC_ID_TCGA_BRCA_mutation_wustl_gene  \\\n",
       "TCGA-A1-A0SB-01               TCGA-A1-A0SB-01A-11D-A142-09   \n",
       "TCGA-A1-A0SD-01               TCGA-A1-A0SD-01A-11D-A10Y-09   \n",
       "TCGA-A1-A0SE-01               TCGA-A1-A0SE-01A-11D-A099-09   \n",
       "TCGA-A1-A0SF-01               TCGA-A1-A0SF-01A-11D-A142-09   \n",
       "TCGA-A1-A0SG-01               TCGA-A1-A0SG-01A-11D-A142-09   \n",
       "\n",
       "                 _GENOMIC_ID_TCGA_BRCA_miRNA_GA  \\\n",
       "TCGA-A1-A0SB-01                             NaN   \n",
       "TCGA-A1-A0SD-01                             NaN   \n",
       "TCGA-A1-A0SE-01                 TCGA-A1-A0SE-01   \n",
       "TCGA-A1-A0SF-01                             NaN   \n",
       "TCGA-A1-A0SG-01                             NaN   \n",
       "\n",
       "                _GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_percentile  \\\n",
       "TCGA-A1-A0SB-01         a2405d64-34eb-4915-abf7-8530151d5cb0   \n",
       "TCGA-A1-A0SD-01         15bad71d-3031-413b-9e8d-6426ae5dfbea   \n",
       "TCGA-A1-A0SE-01         a998e0ce-9248-460f-aabc-2dad452a1ff9   \n",
       "TCGA-A1-A0SF-01         28089e15-5e2c-4e83-ba6c-62b3cb40e431   \n",
       "TCGA-A1-A0SG-01         0df6b948-367d-4951-9d98-d3bebabff63e   \n",
       "\n",
       "                _GENOMIC_ID_data/public/TCGA/BRCA/miRNA_GA_gene  \\\n",
       "TCGA-A1-A0SB-01                                             NaN   \n",
       "TCGA-A1-A0SD-01                                             NaN   \n",
       "TCGA-A1-A0SE-01                                 TCGA-A1-A0SE-01   \n",
       "TCGA-A1-A0SF-01                                             NaN   \n",
       "TCGA-A1-A0SG-01                                             NaN   \n",
       "\n",
       "                _GENOMIC_ID_TCGA_BRCA_gistic2thd  \\\n",
       "TCGA-A1-A0SB-01     TCGA-A1-A0SB-01A-11D-A141-01   \n",
       "TCGA-A1-A0SD-01     TCGA-A1-A0SD-01A-11D-A111-01   \n",
       "TCGA-A1-A0SE-01     TCGA-A1-A0SE-01A-11D-A087-01   \n",
       "TCGA-A1-A0SF-01     TCGA-A1-A0SF-01A-11D-A141-01   \n",
       "TCGA-A1-A0SG-01     TCGA-A1-A0SG-01A-11D-A141-01   \n",
       "\n",
       "                _GENOMIC_ID_data/public/TCGA/BRCA/miRNA_HiSeq_gene  \\\n",
       "TCGA-A1-A0SB-01                                    TCGA-A1-A0SB-01   \n",
       "TCGA-A1-A0SD-01                                    TCGA-A1-A0SD-01   \n",
       "TCGA-A1-A0SE-01                                                NaN   \n",
       "TCGA-A1-A0SF-01                                    TCGA-A1-A0SF-01   \n",
       "TCGA-A1-A0SG-01                                    TCGA-A1-A0SG-01   \n",
       "\n",
       "                 _GENOMIC_ID_TCGA_BRCA_G4502A_07_3  \\\n",
       "TCGA-A1-A0SB-01                                NaN   \n",
       "TCGA-A1-A0SD-01       TCGA-A1-A0SD-01A-11R-A115-07   \n",
       "TCGA-A1-A0SE-01       TCGA-A1-A0SE-01A-11R-A084-07   \n",
       "TCGA-A1-A0SF-01                                NaN   \n",
       "TCGA-A1-A0SG-01                                NaN   \n",
       "\n",
       "                    _GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2  \\\n",
       "TCGA-A1-A0SB-01  a2405d64-34eb-4915-abf7-8530151d5cb0   \n",
       "TCGA-A1-A0SD-01  15bad71d-3031-413b-9e8d-6426ae5dfbea   \n",
       "TCGA-A1-A0SE-01  a998e0ce-9248-460f-aabc-2dad452a1ff9   \n",
       "TCGA-A1-A0SF-01  28089e15-5e2c-4e83-ba6c-62b3cb40e431   \n",
       "TCGA-A1-A0SG-01  0df6b948-367d-4951-9d98-d3bebabff63e   \n",
       "\n",
       "                _GENOMIC_ID_TCGA_BRCA_gistic2 _GENOMIC_ID_TCGA_BRCA_PDMarray  \n",
       "TCGA-A1-A0SB-01  TCGA-A1-A0SB-01A-11D-A141-01                            NaN  \n",
       "TCGA-A1-A0SD-01  TCGA-A1-A0SD-01A-11D-A111-01                TCGA-A1-A0SD-01  \n",
       "TCGA-A1-A0SE-01  TCGA-A1-A0SE-01A-11D-A087-01                TCGA-A1-A0SE-01  \n",
       "TCGA-A1-A0SF-01  TCGA-A1-A0SF-01A-11D-A141-01                            NaN  \n",
       "TCGA-A1-A0SG-01  TCGA-A1-A0SG-01A-11D-A141-01                            NaN  \n",
       "\n",
       "[5 rows x 202 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2=pd.read_csv('BRCA_clinicalMatrix.gz',compression='gzip',sep='\\t',index_col=0)\n",
    "df2.index = pd.Series({x: x.replace('-', '-') for x in df2.index})\n",
    "print(\"Clinical Data Before:\", df2.shape)\n",
    "k=df2.columns.get_loc('PAM50Call_RNAseq')\n",
    "df2=df2[df2.iloc[:,k].isna()==False]\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e78c03fd-abfd-42c0-a2db-3e67c7c332cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data After: (539, 5351)\n",
      "Clinical Data After: (539, 202)\n",
      "CNV and Clinical Data corresponds to same Patients:  True\n",
      "Labels ['Basal' 'Her2' 'LumA' 'LumB' 'Normal'] [ 86  31 279 126  17]\n"
     ]
    }
   ],
   "source": [
    "commonIndexes=np.sort(list(set(df2.index).intersection(set(df.index)))) \n",
    "df=df.loc[commonIndexes] \n",
    "df2=df2.loc[commonIndexes] \n",
    "print(\"Data After:\", df.shape) \n",
    "print(\"Clinical Data After:\", df2.shape) \n",
    "print(\"CNV and Clinical Data corresponds to same Patients: \", False if False in (df.index==df2.index) else True) \n",
    "unique_elements, counts_elements = np.unique(df2.iloc[:,k], return_counts=True) \n",
    "print(\"Labels\", unique_elements, counts_elements)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "682a6675-d9e9-423a-a6b1-b2bc2a75e3c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(539,)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "original_feature_names = df.columns.tolist()\n",
    "X=np.array(df.values, dtype=float)\n",
    "X.shape\n",
    "Y=np.array(df2.values[:,k])\n",
    "Y.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "978a01a9-a906-4d4a-8983-ae3e34710ad3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import StratifiedKFold\n",
    "\n",
    "n_splits = 5\n",
    "skf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8a3c1439-e5c0-4c71-8a4c-f9903793cdb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((432, 5351), (107, 5351))"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "folds = list(skf.split(X, Y))\n",
    "\n",
    "# Select fold 4\n",
    "train_idx, test_idx = folds[4]\n",
    "X_train, X_test = X[train_idx], X[test_idx]\n",
    "Y_train, Y_test = Y[train_idx], Y[test_idx]\n",
    "\n",
    "X_train.shape, X_test.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9494c4ea-d833-4edf-9fd9-d80ca7b91b14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train label distribution:\n",
      " Label\n",
      "LumA      223\n",
      "LumB      101\n",
      "Basal      69\n",
      "Her2       25\n",
      "Normal     14\n",
      "Name: count, dtype: int64\n",
      "Test label distribution:\n",
      " Label\n",
      "LumA      56\n",
      "LumB      25\n",
      "Basal     17\n",
      "Her2       6\n",
      "Normal     3\n",
      "Name: count, dtype: int64\n",
      "Train/test CSVs saved successfully under 'TCGA Dataset Splits/' folder.\n"
     ]
    }
   ],
   "source": [
    "# Convert back to DataFrames for clarity and labeled saving\n",
    "X_train_df = pd.DataFrame(X_train, columns=original_feature_names, index=np.arange(len(X_train)))\n",
    "X_test_df = pd.DataFrame(X_test, columns=original_feature_names, index=np.arange(len(X_test)))\n",
    "Y_train_df = pd.DataFrame(Y_train, columns=[\"Label\"], index=np.arange(len(Y_train)))\n",
    "Y_test_df = pd.DataFrame(Y_test, columns=[\"Label\"], index=np.arange(len(Y_test)))\n",
    "\n",
    "print(\"Train label distribution:\\n\", Y_train_df[\"Label\"].value_counts())\n",
    "print(\"Test label distribution:\\n\", Y_test_df[\"Label\"].value_counts())\n",
    "\n",
    "os.makedirs(\"splits\", exist_ok=True)\n",
    "\n",
    "# Save to CSV files\n",
    "X_train_df.to_csv(\"Datasets/TCGA Dataset Splits/X_train.csv\", index=False)\n",
    "X_test_df.to_csv(\"Datasets/TCGA Dataset Splits/X_test.csv\", index=False)\n",
    "Y_train_df.to_csv(\"Datasets/TCGA Dataset Splits/Y_train.csv\", index=False)\n",
    "Y_test_df.to_csv(\"Datasets/TCGA Dataset Splits/Y_test.csv\", index=False)\n",
    "\n",
    "print(\"Train/test CSVs saved successfully under 'TCGA Dataset Splits/' folder.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a79654f-7c95-4eeb-b107-4587ff9d7ce5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
